File: flut/.flutter/lib/flutter/widgets/focus_manager.dart
import 'package:flutter/material.dart';
import 'package:flut/flut/object.dart';
import 'package:flut/flut/runtime.dart';
import 'package:flut/flutter/foundation/change_notifier.dart';

class _FlutFocusNode extends FocusNode {
  _FlutFocusNode({required FlutRuntime runtime, super.onKeyEvent});
}

class FlutFocusNode extends FlutChangeNotifier<FocusNode> {
  FlutFocusNode.createFromData({
    required super.runtime,
    required super.data,
    required super.target,
  }) : super.createFromData();

  FlutFocusNode.createFromObject({
    required super.runtime,
    required super.oid,
    required super.target,
  }) : super.createFromObject(type: 'FocusNode');

  static FlutFocusNode flutCreate(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return FlutFocusNode.createFromData(
      runtime: runtime,
      data: data,
      target: _FlutFocusNode(
        runtime: runtime,
        onKeyEvent:
            runtime.unpackOptionalCallback(data, 'onKeyEvent')
                as FocusOnKeyEventCallback?,
      ),
    );
  }

  @override
  dynamic getRawProperty(String property) {
    switch (property) {
      case 'hasFocus':
        return flutTarget.hasFocus;
      case 'hasPrimaryFocus':
        return flutTarget.hasPrimaryFocus;
      case 'onKeyEvent':
        return flutTarget.onKeyEvent;
    }
    return super.getRawProperty(property);
  }

  @override
  bool setProperty(String property, dynamic value) {
    switch (property) {
      case 'onKeyEvent':
        flutTarget.onKeyEvent = value as FocusOnKeyEventCallback?;
        return true;
    }
    return super.setProperty(property, value);
  }

  @override
  dynamic callMethod(
    String method,
    List<dynamic> args,
    Map<String, dynamic> kwargs,
  ) {
    switch (method) {
      case 'requestFocus':
        flutTarget.requestFocus();
        return null;
      case 'unfocus':
        flutTarget.unfocus();
        return null;
    }
    return super.callMethod(method, args, kwargs);
  }
}

class FlutFocusScopeNode extends FlutChangeNotifier<FocusScopeNode> {
  FlutFocusScopeNode.createFromData({
    required super.runtime,
    required super.data,
    required super.target,
  }) : super.createFromData();

  FlutFocusScopeNode.createFromObject({
    required super.runtime,
    required super.oid,
    required super.target,
  }) : super.createFromObject(type: 'FocusScopeNode');

  static FlutRealtimeObject flutCreate(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    final node = FocusScopeNode(
      debugLabel: runtime.unpackOptionalField<String>(data, 'debugLabel'),
      onKeyEvent:
          runtime.unpackOptionalCallback(data, 'onKeyEvent')
              as FocusOnKeyEventCallback?,
      skipTraversal: runtime.unpackRequiredField<bool>(data, 'skipTraversal'),
      canRequestFocus: runtime.unpackRequiredField<bool>(
        data,
        'canRequestFocus',
      ),
      traversalEdgeBehavior: runtime.unpackRequiredField<TraversalEdgeBehavior>(
        data,
        'traversalEdgeBehavior',
      ),
      directionalTraversalEdgeBehavior: runtime
          .unpackRequiredField<TraversalEdgeBehavior>(
            data,
            'directionalTraversalEdgeBehavior',
          ),
    );
    return FlutFocusScopeNode.createFromData(
      runtime: runtime,
      data: data,
      target: node,
    );
  }

  @override
  dynamic getRawProperty(String property) {
    switch (property) {
      case 'hasFocus':
        return flutTarget.hasFocus;
      case 'hasPrimaryFocus':
        return flutTarget.hasPrimaryFocus;
      case 'onKeyEvent':
        return flutTarget.onKeyEvent;
    }
    return super.getRawProperty(property);
  }

  @override
  bool setProperty(String property, dynamic value) {
    switch (property) {
      case 'onKeyEvent':
        flutTarget.onKeyEvent = value as FocusOnKeyEventCallback?;
        return true;
    }
    return super.setProperty(property, value);
  }

  @override
  dynamic callMethod(
    String method,
    List<dynamic> args,
    Map<String, dynamic> kwargs,
  ) {
    switch (method) {
      case 'requestFocus':
        flutTarget.requestFocus();
        return null;
      case 'unfocus':
        flutTarget.unfocus();
        return null;
    }
    return super.callMethod(method, args, kwargs);
  }
}

class FlutKeyEventResult extends FlutEnumObject<KeyEventResult> {
  const FlutKeyEventResult()
    : super('KeyEventResult', const {
        'handled': KeyEventResult.handled,
        'ignored': KeyEventResult.ignored,
        'skipRemainingHandlers': KeyEventResult.skipRemainingHandlers,
      });
}

File: flut/flutter/widgets/focus_manager.py
from flut._flut.engine import _flut_pack_value, FlutEnumObject
from flut.flutter.foundation.change_notifier import ChangeNotifier
from flut.flutter.widgets.focus_traversal import TraversalEdgeBehavior


class KeyEventResult(FlutEnumObject):
    handled: "KeyEventResult"
    ignored: "KeyEventResult"
    skipRemainingHandlers: "KeyEventResult"


class FocusNode(ChangeNotifier):
    _flut_type = "FocusNode"

    def __init__(self, onKeyEvent=None):
        self._flut_init_bindings = [
            ("onKeyEvent", onKeyEvent, "FocusOnKeyEventCallback"),
        ]
        super().__init__()

    @property
    def onKeyEvent(self):
        return self._flut_get("onKeyEvent")

    @onKeyEvent.setter
    def onKeyEvent(self, callback):
        self._flut_set("onKeyEvent", callback, "FocusOnKeyEventCallback")

    @property
    def hasFocus(self):
        return self._flut_get("hasFocus")

    @property
    def hasPrimaryFocus(self):
        return self._flut_get("hasPrimaryFocus")

    def dispose(self):
        self._flut_call("dispose")

    def requestFocus(self):
        self._flut_call("requestFocus")

    def unfocus(self):
        self._flut_call("unfocus")


class FocusScopeNode(FocusNode):
    _flut_type = "FocusScopeNode"

    def __init__(
        self,
        *,
        onKeyEvent=None,
        debugLabel=None,
        skipTraversal=False,
        canRequestFocus=True,
        traversalEdgeBehavior=TraversalEdgeBehavior.closedLoop,
        directionalTraversalEdgeBehavior=TraversalEdgeBehavior.stop,
    ):
        props = {}
        if debugLabel is not None:
            props["debugLabel"] = debugLabel
        props["skipTraversal"] = skipTraversal
        props["canRequestFocus"] = canRequestFocus
        props["traversalEdgeBehavior"] = _flut_pack_value(traversalEdgeBehavior)
        props["directionalTraversalEdgeBehavior"] = _flut_pack_value(
            directionalTraversalEdgeBehavior
        )
        self._flut_init_props = props
        self._flut_init_bindings = [
            ("onKeyEvent", onKeyEvent, "FocusOnKeyEventCallback"),
        ]
        # Skip FocusNode.__init__ which would set its own bindings.
        ChangeNotifier.__init__(self)
